File: bindings/java/src/main/java/com/hyperpolymath/proven/ProvenIds.java
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// IDS protocol types for proven-servers.

package com.hyperpolymath.proven;

/**
 * IDS protocol types for proven-servers matching the Idris2 ABI tags.
 * @author Jonathan D.A. Jewell
 */
public final class ProvenIds {
    private ProvenIds() {}

    /** AlertSeverity (tags 0-3). */
    public enum AlertSeverity {
        LOW(0),
        MEDIUM(1),
        HIGH(2),
        CRITICAL(3);

        private final int tag;
        AlertSeverity(int tag) { this.tag = tag; }
        public int tag() { return tag; }

        public static AlertSeverity fromTag(int tag) {
            for (AlertSeverity v : values()) if (v.tag == tag) return v;
            return null;
        }
    }

    /** DetectionMethod (tags 0-3). */
    public enum DetectionMethod {
        SIGNATURE(0),
        ANOMALY(1),
        STATEFUL(2),
        HEURISTIC(3);

        private final int tag;
        DetectionMethod(int tag) { this.tag = tag; }
        public int tag() { return tag; }

        public static DetectionMethod fromTag(int tag) {
            for (DetectionMethod v : values()) if (v.tag == tag) return v;
            return null;
        }
    }

    /** IdsProtocol (tags 0-6). */
    public enum IdsProtocol {
        TCP(0),
        UDP(1),
        ICMP(2),
        DNS(3),
        HTTP(4),
        TLS(5),
        SSH(6);

        private final int tag;
        IdsProtocol(int tag) { this.tag = tag; }
        public int tag() { return tag; }

        public static IdsProtocol fromTag(int tag) {
            for (IdsProtocol v : values()) if (v.tag == tag) return v;
            return null;
        }
    }

    /** IdsAction (tags 0-4). */
    public enum IdsAction {
        ALERT(0),
        DROP(1),
        LOG(2),
        BLOCK(3),
        PASS(4);

        private final int tag;
        IdsAction(int tag) { this.tag = tag; }
        public int tag() { return tag; }

        public static IdsAction fromTag(int tag) {
            for (IdsAction v : values()) if (v.tag == tag) return v;
            return null;
        }
    }

    /** Direction (tags 0-2). */
    public enum Direction {
        INBOUND(0),
        OUTBOUND(1),
        BOTH(2);

        private final int tag;
        Direction(int tag) { this.tag = tag; }
        public int tag() { return tag; }

        public static Direction fromTag(int tag) {
            for (Direction v : values()) if (v.tag == tag) return v;
            return null;
        }
    }

    /** ThreatLevel (tags 0-4). */
    public enum ThreatLevel {
        INFO(0),
        LOW(1),
        MEDIUM(2),
        HIGH(3),
        CRITICAL(4);

        private final int tag;
        ThreatLevel(int tag) { this.tag = tag; }
        public int tag() { return tag; }

        public static ThreatLevel fromTag(int tag) {
            for (ThreatLevel v : values()) if (v.tag == tag) return v;
            return null;
        }
    }
}

File: bindings/kotlin/src/main/kotlin/com/hyperpolymath/proven/ProvenIds.kt
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// IDS protocol types for proven-servers.

package com.hyperpolymath.proven

/** AlertSeverity matching the Idris2 ABI tags. */
enum class AlertSeverity(val tag: Int) {
    ALERT_SEVERITY__LOW(0),
    ALERT_SEVERITY__MEDIUM(1),
    ALERT_SEVERITY__HIGH(2),
    ALERT_SEVERITY__CRITICAL(3);

    companion object {
        fun fromTag(tag: Int): AlertSeverity? = entries.find { it.tag == tag }
    }
}

/** DetectionMethod matching the Idris2 ABI tags. */
enum class DetectionMethod(val tag: Int) {
    SIGNATURE(0),
    ANOMALY(1),
    STATEFUL(2),
    HEURISTIC(3);

    companion object {
        fun fromTag(tag: Int): DetectionMethod? = entries.find { it.tag == tag }
    }
}

/** IdsProtocol matching the Idris2 ABI tags. */
enum class IdsProtocol(val tag: Int) {
    TCP(0),
    UDP(1),
    ICMP(2),
    DNS(3),
    HTTP(4),
    TLS(5),
    SSH(6);

    companion object {
        fun fromTag(tag: Int): IdsProtocol? = entries.find { it.tag == tag }
    }
}

/** IdsAction matching the Idris2 ABI tags. */
enum class IdsAction(val tag: Int) {
    ALERT(0),
    DROP(1),
    LOG(2),
    BLOCK(3),
    PASS(4);

    companion object {
        fun fromTag(tag: Int): IdsAction? = entries.find { it.tag == tag }
    }
}

/** Direction matching the Idris2 ABI tags. */
enum class Direction(val tag: Int) {
    INBOUND(0),
    OUTBOUND(1),
    BOTH(2);

    companion object {
        fun fromTag(tag: Int): Direction? = entries.find { it.tag == tag }
    }
}

/** ThreatLevel matching the Idris2 ABI tags. */
enum class ThreatLevel(val tag: Int) {
    INFO(0),
    THREAT_LEVEL__LOW(1),
    THREAT_LEVEL__MEDIUM(2),
    THREAT_LEVEL__HIGH(3),
    THREAT_LEVEL__CRITICAL(4);

    companion object {
        fun fromTag(tag: Int): ThreatLevel? = entries.find { it.tag == tag }
    }
}
